gossip/src/duplicate_shred_handler.rs

```diff
@@ -34,7 +34,7 @@ pub struct DuplicateShredHandler {
     // together the chunks into the original proof before anything useful is done.
     buffer: HashMap<(Slot, Pubkey), BufferEntry>,
     // Slots for which a duplicate proof is already ingested.
-    consumed: HashSet<Slot>,
+    consumed: HashMap<Slot, bool>,
```

Comment:
So, has_duplicate_shreds_in_slot is an one-way street, once it's duplicate it will never go the other way. Under normal cases the chunks of duplicate proof should arrive consecutively, so the slot quickly goes into consumed. So we could also not check has_duplicate_shreds_in_slot at all when we ingest the chunks, only check that when we need to prune the cache or when we pieced together a full proof.

So consumed can still be a HashSet, slots go in there either when a proof is accepted through Gossip or when has_duplicate_shreds_in_slot is true, because I think a false value in consumed currently only tells us has_duplicate_shreds_in_slot was false when we first inserted the chunks, we probably don't care about getting more chunks even though has_duplicate_shreds_in_slot has silently turned true, we will find out soon when the full proof is accepted.

Reply:
> Under normal cases the chunks of duplicate proof should arrive consecutively, so the slot quickly goes into consumed.

This is subject to network conditions, gossip propagation, spams from bad actors, etc; and we cannot rely on that.
`true/false` here ensures that `has_duplicate_shreds_in_slot` is invoked at most twice on each slot which is good enough.

Reply:
I mean, right now the behavior is:
On the first chunk read has_duplicate_shreds_in_slot once, if this value is false, then you would happily ingest chunks for this slot until a full proof is accepted, so the only difference between this change and not checking has_duplicate_shreds_in_slot on consumption is that you are protected once when the first chunk arrives. Not sure how much use that is.

Looking back, consumed is really a per slot status, if it gets even more complicated maybe we need to rethink whether flatten slot -> pubkey  hashmap is better, because we did lose the ability to put in isolation between slots.

That said, I could approve this change because I don't expect consumed to become even more complicated soon.

Reply:
> so the only difference between this change and not checking has_duplicate_shreds_in_slot on consumption is that you are protected once when the first chunk arrives. Not sure how much use that is.

You may receive chunks from many different nodes **for the same slot** which already has a duplicate proof in blockstore. With this commit `self.consumed` will skip them **all** with **only a single** call on blockstore.

Reply:
I mean:

Since we only check has_duplicate_shreds_in_slot once when the first chunk arrives and nothing after that, you only skip chunks for this slot when has_duplicate_shreds_in_slot is true at that time, if has_duplicate_shreds_in_slot is false when the first chunk arrives, then your behavior is the same as not checking has_duplicate_shreds_in_slot at all.

So you are telling me under many cases when the first chunk arrives has_duplicate_shreds_in_slot is probably true already, okay then.

Reply:
If `has_duplicate_shreds_in_slot` was initially `false`, and it becomes `true` by this handler itself then immediately the `slot` is also set to `true` in `self.consumed`:
https://github.com/solana-labs/solana/blob/ceb225f36/gossip/src/duplicate_shred_handler.rs#L128-L135
and therefore skipping all the future chunks received for this same slot; 

If a duplicate proof is pushed to blockstore here:
https://github.com/solana-labs/solana/blob/ceb225f36/core/src/window_service.rs#L145-L150
it is also pushing duplicate shreds to gossip which is then ingested in this handler which then again sets the slot to `true` in `self.consumed`.

Reply:
I think there is no discrepancy here. Under both cases you mentioned, the handler will see a complete duplicate proof so under the old code path, the slot is going into consumed hashset, and all partial proofs and future proofs will be dumped. So a true value in consumed is the same as the slot in the original consumed hashset, and a false value is not doing much other than saying do not query has_duplicate_shreds_in_slot again.